cfep-09.md

```diff
@@ -0,0 +1,99 @@
+
+<table>
+<tr><td> Pinning Proposal System for Automatic Rebuilds</td><td> Pinning Proposal </td>
+<tr><td> Status </td><td> Proposed </td></tr>
+<tr><td> Author(s) </td><td> Christopher J. "CJ" Wright &lt;cjwright4242@gmail.com&gt;; Matthew R. Becker &lt;becker.mr@gmail.com&gt;</td></tr>
+<tr><td> Created </td><td> Feb 20, 2019</td></tr>
+<tr><td> Updated </td><td> Aug 30, 2016</td></tr>
+<tr><td> Discussion </td><td> NA </td></tr>
+<tr><td> Implementation </td><td> NA </td></tr>
+</table>
+
+## Abstract
+
+This CFEP proposes a change to how our internal pinning system works.
+The new system will comprise of a new repo for proposed migrations.
+To propose a migration a PR will be issued with a new version of
+the global pinning file and some metadata about the migration needed 
+(if any).
+Once the PR is squashed and merged the bot will take the metadata
+to calculate which packages need to be rebuilt and start issuing
+PRs with the new pinning as a local pinning file to the feedstocks
+in topo order.
+
+This CFEP is taken from @beckermr's work on [conda-forge.github.io](https://github.com/conda-forge/conda-forge.github.io/issues/712).
+
+## Motivation
+Currently when we move pinnings we move the pin in the ``conda-forge-pinning-feedstock``.
+Then we start a migration using the ``regro-cf-autotick-bot``.
+This order of operations causes some issues, when new feedstocks
+are created they recieve the new pinnings even if their dependencies
+are not rebuilt with the new pinnings themselves.
+
+## Specification
+
+### The migrations proceed as follows:
+1. We make a PR on the `migrations` repo with a change to the global pinnings file, generating a new version. 
+2. Once the change is merged, the bot picks it up and starts going through the graph in topo order. 
+3. A migration PR is issued for a node only if 
+    1. The node depends on the changed pinnings.
+    2. It has no dependencies that depend on the new pinnings and have not been migrated.
+    3. It is not currently being migrated by an earlier migration. Per @scopatz, this state is stored in `libcflib` with a REST API.
```

Comment:
This could cause issues of its own. One could imagine a very difficult migration (eg compiler major ABI breaks) which comes before a very simple migration. In this case we may have to wait a long time for bottleneck nodes to get through with the hard change before the simple change is implemented. This is especially painful if the simple migration is a security fix where speed is of the essence.

I don't know if the `REST` detail is needed in this point. If the bot has started migrating things, I imagine it could figure out which migrations it has open and who has claim to which nodes and in what order. The `REST` API could be important to how the bot figures out which migrations it needs to run.

Reply:
Indeed, we could have issues around this. OTOH, if we are generating more than one migration PR per feedstock, or if successive changes in the global pinnings are incompatible, then we might have other issues. I'd say this is the least worst solution. 

We could add a mechanism to force a second migration to proceed, but we would need to develop a policy around what to do with feedstock PRs that are already open from the previous migration. We would also have to dump the time ordered versioning scheme as well to make overrides like this work.

Reply:
I would also add that for very intrusive/big migrations, we would want to avoid this automated system anyways in favor of work by hand. For example, if we need to do another compiler migration of every feedstock, I would argue we would want a separate set of labels for that in order to rebuild nearly everything before we release them. So I think one of the guidelines we should add here is that migrations that fall roughly under the "To Big to Fail" policy should not  be done via the global pinnings file.

Reply:
> if successive changes in the global pinnings are incompatible, then we might have other issues.

I'm not certain what you mean by this. I think one of our central assumptions is that any given version of the pinning are self consistent.

Reply:
I mean successive commits to the repo. We can't arbitrarily reorder git commits on the global pinnings file. Thus letting migrations hop ahead of one another might not generate the correct pinnings file.